### ISSUE TITLE
AbsentLink doesn't work if the pattern contains a disconnected mandatory clause

### ISSUE THREAD

**Issue by leungmanhin** [opencog/atomspace#288]
For example:
```
(clear)
(ListLink (ConceptNode "A") (ConceptNode "B"))
(ListLink (ConceptNode "C") (ConceptNode "D"))

(define test
    (BindLink
        (VariableList
            (TypedVariableLink
                (VariableNode "$var1")
                (TypeNode "ConceptNode")
            )
            (TypedVariableLink
                (VariableNode "$var2")
                (TypeNode "ConceptNode")
            )
            (TypedVariableLink
                (VariableNode "$var3")
                (TypeNode "ConceptNode")
            )
        )
        (AndLink
            (ListLink
                (VariableNode "$var1")
                (VariableNode "$var2")
            )
            (AbsentLink
                (ListLink
                    (VariableNode "$var3")
                    (ConceptNode "X")
                )
            )
        )
        (ListLink
            (VariableNode "$var1")
            (VariableNode "$var2")
        )
    )
)

(cog-bind test)
```
The above will not return the two ListLinks that was created at the beginning.

However it works if the pattern contains only AbsentLinks because that is specifically handled here:
https://github.com/opencog/atomspace/blob/master/opencog/query/Implicator.cc#L105-L125

It kind of work if the AbsentLinks are connected to some mandatory clause in the pattern, for example:
```
(clear)
(ListLink (ConceptNode "A") (ConceptNode "B"))
(ListLink (ConceptNode "C") (ConceptNode "D"))

(define test2
    (BindLink
        (VariableList
            (TypedVariableLink
                (VariableNode "$var1")
                (TypeNode "ConceptNode")
            )
            (TypedVariableLink
                (VariableNode "$var2")
                (TypeNode "ConceptNode")
            )
        )
        (AndLink
            (ListLink
                (VariableNode "$var1")
                (VariableNode "$var2")
            )
            (AbsentLink
                (ListLink
                    (VariableNode "$var1")
                    (ConceptNode "D")
                )
            )
        )
        (ListLink
            (VariableNode "$var1")
            (VariableNode "$var2")
        )
    )
)

(cog-bind test2)
```
will return `(ListLink (ConceptNode "A") (ConceptNode "B"))`. It actually confuses me as I was expecting the present of AbsentLink will make the pattern matcher to evaluate all possible solutions [in this example, they are `(ListLink (ConceptNode "A") (ConceptNode "B"))` and `(ListLink (ConceptNode "C") (ConceptNode "D"))`] and reject the pattern if any of them [which is `(ListLink (ConceptNode "C") (ConceptNode "D"))`] doesn't satisfy it, and hence returns nothing as a result, but I am not sure if it is the intended behavior...

**Comment by linas** [comment]
Yep, the first example illustrates a bug.

For the second case, think of "absent" meaning "not present in the atomspace". In the second example, there are two ways to ground (v1, v2) -- it can be (A,B) or (C,D)  In the first case, the absent-link would be (A,D) and that is indeed not present in the atomspace, so (A,B) is an acceptable grounding.  For the second case, the absentlink is (C,D), but that is present in the atomspace, so that grounding is rejected.  

The absent link was originally developed for rejecting matches to certain sentences. For example, say you want to match the sentence "I have an apple", but only if there is no modifier to "apple", e.g. to reject the match to "I have a green apple"  This is useful for determining meaning: for example, in the first sentence, "apple" might refer to "iphone"; but its very unlikely that "green apple" would be an iphone.

**Comment by linas** [comment]
this is fixed, now, right?

**Comment by leungmanhin** [comment]
Yes fixed by #323
